### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by qle2**
This PR was created to address Issues #74. In it, I: 

- Mainly revamped our unit tests in the `tests`. 

- Redid the config.yml so that the key CircleCI generates in every test is contained in each branch. This is to contain CircleCI errors from going to other branches.

- Removed rdkit2d-SMILES relationships and reduce the number of fragments we're generating for our Neo4j graphs.

I will be adding documentations to each testing functions soon

### CODE REVIEW

File: tests/test_ingest.py [hunk 1]
```diff
@@ -1,62 +1,16 @@
...
+import pandas as pd
+
+
+@pytest.mark.parametrize('algorithm, data, exp, tuned, delete, directory', [('rf', 'Lipo-short.csv', 'exp', False,
+                                                                             False, True)])
+def test_ingest(__model_object__):
+    model1 = __model_object__
+    assert type(model1.data) is pd.core.frame.DataFrame
+    assert type(model1.smiles_series) is pd.Series
+    assert len(model1.data.columns) == 2
```
Comment (Dlux804) [inline tests/test_ingest.py:16]: How are you going to handle other situations, such as multi-label classifications? Something to consider with @dickeygh

File: core/neo4j/.idea/neo4j.iml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+  <component name="PyDocumentationSettings">
+    <option name="format" value="PLAIN" />
+    <option name="myDocStringFormat" value="Plain" />
+  </component>
+  <component name="TestRunnerService">
+    <option name="PROJECT_TEST_RUNNER" value="pytest" />
+  </component>
+</module>
```
[the remark this answers is not in the document]
Comment (qle2) [reply core/neo4j/.idea/neo4j.iml:15]: It was automatically rendered. I'm not sure what it is or what triggered it.

File: tests/model_fixture.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            model1 = models.MlModel(algorithm=algorithm, dataset=data, target=exp, tune=tuned, feat_meth=[0], cv=2,
+                                    opt_iter=2)
+    else:
+        model1 = models.MlModel(algorithm=algorithm, dataset=data, target=exp, tune=tuned, feat_meth=[0], cv=2,
+                                opt_iter=2)
+    model1.featurize()
+    model1.data_split(val=0.1)
+    model1.reg()
+    model1.run()
+    model1.analyze()
```
[the remark this answers is not in the document]
Comment (qle2) [reply tests/model_fixture.py:62]: I just thought of the same idea and I am changing my fixture to load a pickled model.

File: tests/test_ingest.py (hunk 1 above)
Comment (qle2) [reply tests/test_ingest.py:16]: Will do. I built this with only regression in mind.

File: tests/test_regressors.py [hunk 4]
```diff
@@ -1,70 +1,57 @@
...
+@pytest.mark.parametrize('algorithm, data, exp, tuned, directory', [('nn', 'Lipo-short.csv', 'exp', False, True)])
+def test_nn_get_regressor(__data_split_model__):
+    model1 = __data_split_model__
+    model1.get_regressor()
+    assert model1.regressor, "Can't call for regressor object"
+    assert type(model1.fit_params) is dict, "fit params is not a dictionary"
+    assert model1.task_type == 'regression'
+
+
+sklearn_list = ['ada', 'svr', 'rf', 'gdb', 'mlp', 'knn']
```
[the remark this answers is not in the document]
Comment (qle2) [reply tests/test_regressors.py:23]: Yes, `pytest.mark.parametrize` will loop through the list of algorithms. I only use it for parts of `core` where algorithms are involved since they are more time intensive.

File: tests/test_storage.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+@pytest.mark.parametrize('algorithm, data, exp, tuned, delete, directory', [('rf', 'Lipo-short.csv', 'exp', False,
+                                                                             True, True)])
+def test_store(__model_object__):
+    """"""
+    model1 = __model_object__
+    model1.store()
+    assert os.path.isfile(''.join([model1.run_name, '_data.csv']))
+    assert os.path.isfile(''.join([model1.run_name, '_attributes.json']))
+    delete_files(model1.run_name)
+
```
[the remark this answers is not in the document]
Comment (qle2) [reply tests/test_storage.py:24]: I created the `delete_files` function mostly because I ran these tests on pycharm. I'm not sure if files are created when CircIeC runs them. Will test it out

File: main.py [hunk 6]
```diff
@@ -150,11 +151,12 @@ def single_model():
...
-        if model1.algorithm != 'nn':  # issues pickling NN models
-            model1.pickle_model()
-        model1.store()
-        model1.org_files(zip_only=True)
+        # model1.analyze()
+        # if model1.algorithm != 'nn':  # issues pickling NN models
+        #     model1.pickle_model()
+        #
+        # model1.store()
+        # model1.org_files(zip_only=True)
```
[the remark this answers is not in the document]
Comment (qle2) [reply main.py:159]: I did it for some quick testing when running `single_model`. Will uncomment them.

File: tests/model_fixture.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (qle2) [reply tests/model_fixture.py:62]: Scratch that. I think loading pickled model kind of defeats the purpose of testing. While we can still make tests for their content, the function themselves need to be run for a complete test

File: tests/test_regressors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dlux804) [reply tests/test_regressors.py:23]: Is this costly in terms of time it takes to run the tests?

File: tests/model_fixture.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dlux804) [reply tests/model_fixture.py:62]: My thought is that you can load only part of the pickled model.  Or we have a model pickled at each point in the pipeline.  For example, if we want to test `data_split()` we can load the just the featurized data from a pickle model then run it.  Or for `analyze()` we load a pre-tuned, pre-trained model and do the analysis on that.  The goal is to recreate the input to the function without having to regenerate it *ab initio*

File: tests/test_regressors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (qle2) [reply tests/test_regressors.py:23]: For `test_get_regressor`, since we're just grabbing the correct algorithm for this specific test, it does not take much time at all. But for `test_hypertune`, it does take a couple of seconds longer to Keras model. However, since I'm using a test data with only 10 molecules, it does not take much time at all.

File: core/neo4j/nodes_to_neo4j.py [hunk 7]
```diff
@@ -154,25 +154,25 @@ def nodes(self):
...
+    #         try:
+    #             df_rdkit2d_features = df_rdkit2d_features.rename(columns={self.target_name: 'target'})
+    #         except KeyError:
+    #             pass
+    #         df_rdkit2d_features = df_rdkit2d_features.drop('target', axis=1)
+    #         df_rdkit2d_features.apply(__merge_molecules_and_rdkit2d__, g=g, axis=1)
+    #         t3 = time.perf_counter()
+    #         print(f"Finished creating nodes and relationships between SMILES and rdkit2d features in {t3 - t2}sec")
+    # else:
+    #     pass
```
[the remark this answers is not in the document]
Comment (qle2) [reply core/neo4j/nodes_to_neo4j.py:175]: Do you also one to keep the ones that rank high on feature_importance such as qed, TPSA, partial charges...?

File: core/neo4j/nodes_to_neo4j.py [hunk 8]
```diff
@@ -154,25 +154,20 @@ def nodes(self):
...
+        if len(list(record)) > 0:  # Already created rdkit2d feature for this dataset
+            print(f"This dataset, {self.dataset}, and its rdkit2d features already exist in the database. Moving on")
+        else:  # If not
+            print("Creating rdki2d features")
+            df_rdkit2d_features = self.data.filter(regex='smiles|fr_|Count|Num|Charge', axis=1)
+            df_rdkit2d_features.apply(__merge_molecules_and_rdkit2d__, g=g, axis=1)
+            t3 = time.perf_counter()
+            print(f"Finished creating nodes and relationships between SMILES and rdkit2d features in {t3 - t2}sec")
+    else:
+        pass
```
Comment (Dlux804) [inline core/neo4j/nodes_to_neo4j.py]: Is this the code block that links molecules to feature nodes in neo4j?

File: core/neo4j/output_to_neo4j.py [hunk 9]
```diff
@@ -83,5 +86,6 @@ def output_to_neo4j(port, username, password):
         prep = prep_from_outputs.Prep(df_from_attributes, df_from_predictions, df_from_data)
         prep.to_neo4j(port=port, username=username, password=password)
 
+
 # You can uncomment this to run the process of importing data from output files into Neo4j based on our ontology
-# output_to_neo4j()
+# output_to_neo4j(port="bolt://localhost:7687", username="neo4j", password="password")
```
Comment (Dlux804) [inline core/neo4j/output_to_neo4j.py:91]: last we spoke, you said password needed to be 1234.  Should this line be changed depending on what the Neo4j server is using?

File: core/neo4j/prep_from_outputs.py [hunk 10]
```diff
@@ -96,4 +92,4 @@ def to_neo4j(self, port, username, password):
         Graph(self.neo4j_params["port"], username=self.neo4j_params["username"],
               password=self.neo4j_params["password"])  # Test connection to Neo4j
         nodes_to_neo4j.nodes(self)
-        rel_to_neo4j.relationships(self)
+        # rel_to_neo4j.relationships(self)
```
Comment (Dlux804) [inline core/neo4j/prep_from_outputs.py:95]: What is this changing?

File: new_tests/test_ingest.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+from core.models import rds, cds  # regression and classification data list
+files, directory = file_list('pickle/init')
+
+
+@pytest.mark.parametrize('pkl', files)
+@pytest.mark.parametrize('directory', [directory])
+def test_load_smiles(__unpickle_model__):
+    """
+    Objective: Test ingest.py's load_smiles. Check if daa instance is a dataframe, smiles_series instance is a
+                pandas Series object and that we have the correct amount of columns depending on the dataset.
```
Comment (Dlux804) [inline new_tests/test_ingest.py:16]: Typo? "Check if daa instance is dataframe..."

File: new_tests/test_keras.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    model1 = __run_all__
+    delete_files(model1.run_name)
+
+
+@pytest.mark.parametrize('algorithm, data, exp, tuned, directory', [('nn', 'Lipo-short.csv', 'exp', True, True)])
+def test_nn_tuned(__run_all__):
+    """
+    Objective: Run a tuned keras model from start to get prediction results.
+    Note: Since I can't unload a h5 file to get important information for almost all tests, I decided to test its
+            functionality by running a keras model from start to getting prediction results
```
Comment (Dlux804) [inline new_tests/test_keras.py:28]: Why can't you unload an .h5?

File: core/neo4j/nodes_to_neo4j.py (hunk 8 above)
Comment (qle2) [reply core/neo4j/nodes_to_neo4j.py]: Yes. I only kept the features that are counts and fragments such as fr_Br. If I remember correctly, it was reduced to a little over 100 from 187 features.

File: core/neo4j/output_to_neo4j.py (hunk 9 above)
Comment (qle2) [reply core/neo4j/output_to_neo4j.py:91]: `password` is the default password for Neo4j. It seems like if you name your database's password as `password`, you can just connect to the database without specifying anything since the default port is `bolt://localhost:7687` and default username is `neo4j`. However, CircleCI threw a fit every time they are not specified.

File: core/neo4j/prep_from_outputs.py (hunk 10 above)
Comment (qle2) [reply core/neo4j/prep_from_outputs.py:95]: It was commented out when I was testing my code. Will uncomment it

File: new_tests/test_ingest.py (hunk 11 above)
Comment (qle2) [reply new_tests/test_ingest.py:16]: Fixed

File: new_tests/test_keras.py (hunk 12 above)
Comment (qle2) [reply new_tests/test_keras.py:28]: I can unload it to fit and predict but I haven't found a way to extract any other information like I could with a pkl file.
